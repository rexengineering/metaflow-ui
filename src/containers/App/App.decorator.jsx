import {ThemeProvider} from "@material-ui/core/styles";
import React from "react";
import buildTheme from "../../theme";
import PropTypes from "prop-types";
import App from "./App";

function AppDecorator({ keyValue }){
    const theme = buildTheme();
    return (
        <ThemeProvider key={keyValue} theme={theme}>
            <App />
        </ThemeProvider>
    )
}

AppDecorator.propTypes= {
    keyValue: PropTypes.string.isRequired,
}

export default AppDecorator;