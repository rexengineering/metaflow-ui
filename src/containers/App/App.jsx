import React, {useEffect} from "react";
import {
  CssBaseline,
  makeStyles,
} from "@material-ui/core";
import clsx from "clsx";
import Debugging from "../../components/Debugging";
import {connect} from "react-redux";
import fetchAvailableWorkflows  from "../../store/thunks/fetchAvailableWorkflows";
import eventBroadcast from "../../store/thunks/subscription/eventBroadcast";
import Workflow from "../../components/Workflow";
import keepAlive from "../../store/thunks/subscription/keepAlive";

const useStyles = makeStyles((theme) => ({
  app: {
    display: "flex",
    height: "100vh",
  },
  pane: {
    width: "100%",
    background: theme.palette.common.white,
    color: theme.palette.text.secondary,
    display: "flex",
  },
  tray: {
    width: "100%",
    background: theme.palette.background.default,
  },
  tray1: {
    width: "25%",
    flexShrink: 0,
    backgroundColor: theme.palette.grey[800],
    padding: theme.spacing(2, 3, 6),
    display: "flex",
    flexDirection: "column",
    justifyContent: "space-between",
    overflow: "auto",
  },
  tray2: {
    padding: theme.spacing(3),
    overflow: "auto",
    transition: theme.transitions.create('margin', {
      easing: theme.transitions.easing.easeOut,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
}));

function App({ instantiatedWorkflows, getAvailableWorkflows, keepAliveSubscription, dispatch }) {
  const classes = useStyles();

  useEffect(() => getAvailableWorkflows(), []);
  useEffect(() => eventBroadcast(dispatch), []);
  useEffect(() => {
    const interval = setInterval(() => keepAliveSubscription(), 2000);
    return () => clearInterval(interval);
  }, []);

  return (
    <div className={classes.app}>
      <CssBaseline />
      <section className={classes.pane}>
        <section className={clsx(classes.tray, classes.tray1)}>
          <Debugging />
        </section>
        <section className={clsx(classes.tray, classes.tray2)} data-testid="tray2">
          {
            Array.isArray(instantiatedWorkflows) && instantiatedWorkflows.map(({ iid }) => (
                <Workflow identifier={iid} key={iid} />
            ))
          }
        </section>
      </section>
    </div>
  );
}

const mapStateToProps = ({ rexFlow: { instantiatedWorkflows }}) => ({
  instantiatedWorkflows
});

const mapDispatchToProps = (dispatch) => ({
  getAvailableWorkflows: () => dispatch(fetchAvailableWorkflows()),
  keepAliveSubscription: () => dispatch(keepAlive()),
  dispatch,
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
