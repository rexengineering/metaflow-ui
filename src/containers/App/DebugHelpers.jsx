import React from "react";
import { Button,  makeStyles, Typography } from "@material-ui/core";
import WorkflowInstantiator from "../../components/WorkflowInstantiator";
import { fetchTasks } from "../../store/thunks/thunks";
import { useDispatch } from "react-redux";

const useStyles = makeStyles((theme) => ({
  root: {
    padding: theme.spacing(1),
  },
  deployment: {
    marginBottom: theme.spacing(2),
  },
  deploymentTitle: {
    marginBottom: theme.spacing(1),
  },
  button: {
    marginTop: theme.spacing(2),
    display: "block",
  },
}));

function DebugHelpers({ deployments, isAutomaticState, setIsAutomaticState, instanceID }) {
  const classes = useStyles();
  const isDeploymentsAvailable = Array.isArray(deployments) && !!deployments.length;
  const dispatch = useDispatch();

  return (
    <div className={classes.root}>
      <Typography variant="h5" align="center">
        Initiate Workflows
      </Typography>
      {!isDeploymentsAvailable && (
        <Typography>There are no workflows available</Typography>
      )}
      {Array.isArray(deployments) &&
        deployments.map((deploymentID) => (
          <div key={deploymentID} className={classes.deployment} data-testid="deployment">
            <Typography
              variant="body2"
              className={classes.deploymentTitle}
            >
              {deploymentID}
            </Typography>
            <WorkflowInstantiator deploymentID={deploymentID} instanceID={instanceID} />
          </div>
        ))}
      <Typography variant="h5" align="center">
        State Helpers
      </Typography>
      <Button
        color="secondary"
        type="button"
        variant="contained"
        onClick={() => dispatch(fetchTasks())}
        className={classes.button}
        data-testid="update-state-button"
      >
        Update state
      </Button>
      <Button
        variant="contained"
        type="button"
        onClick={(currentState) => setIsAutomaticState(!currentState)}
        className={classes.button}
        data-testid="toggle-state-button"
      >
        Toggle automatic state
      </Button>
      <Typography>
        Automatic state fetching is {isAutomaticState ? "On" : "Off"}
      </Typography>
    </div>
  );
}

export default DebugHelpers;
