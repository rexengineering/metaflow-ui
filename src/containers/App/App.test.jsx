import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import App, { MISC_DRAWER_WIDTH } from "./App";
import userEvent from "@testing-library/user-event";

jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
  useDispatch: () => jest.fn(),
}));

describe("<App />", () => {
  it("should render", () => {
    const { asFragment } = render(<App />);
    const initialRender = asFragment();

    expect(initialRender).not.toBeNull();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(<App />);
    const initialRender = asFragment();

    expect(initialRender).toMatchSnapshot();
  });

  it("should start with the misc drawer closed" , () => {
    render(<App />);
    const drawer = screen.getByTestId("misc-drawer").querySelector(".MuiDrawer-paper");
    expect(drawer).toHaveStyle("visibility: hidden");
  });


  it("should toggle the misc drawer when its button is pressed" , async () => {
    render(<App />);
    const drawer = screen.getByTestId("misc-drawer").querySelector(".MuiDrawer-paper");
    const drawerToggleButton = screen.getByTestId("drawer-toggle-button");

    expect(drawer).not.toHaveStyle("transform: none");
    await waitFor(() => {
      userEvent.click(drawerToggleButton);
    });
    expect(drawer).toHaveStyle("transform: none");
    await waitFor(() => {
      userEvent.click(drawerToggleButton);
    });
    expect(drawer).not.toHaveStyle("transform: none");;
  });


  it("should resize the content based on drawer state" , async () => {
    render(<App />);
    const drawer = screen.getByTestId("misc-drawer").querySelector(".MuiDrawer-paper");
    const drawerToggleButton = screen.getByTestId("drawer-toggle-button");
    const content = screen.getByTestId("tray2");

    expect(drawer).toHaveStyle(`width: ${MISC_DRAWER_WIDTH}px`);
    expect(content).not.toHaveStyle(`margin-right: ${MISC_DRAWER_WIDTH}px`);

    await waitFor(() => {
      userEvent.click(drawerToggleButton);
    });
    expect(content).toHaveStyle(`margin-right: ${MISC_DRAWER_WIDTH}px`);

    await waitFor(() => {
      userEvent.click(drawerToggleButton);
    });
    expect(content).not.toHaveStyle(`margin-right: ${MISC_DRAWER_WIDTH}px`);
  });
});
