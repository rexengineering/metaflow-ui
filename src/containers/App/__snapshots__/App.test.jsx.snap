// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-app-10"
  >
    <section
      class="makeStyles-container-17 makeStyles-panel-16"
    >
      <section
        class="makeStyles-container-17 makeStyles-panel-18 makeStyles-tray-11"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-14 MuiButton-containedSecondary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Update state
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-14"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Toggle automatic state
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Automatic state fetching is Off
        </p>
      </section>
      <section
        class="makeStyles-container-17 makeStyles-panel-18 makeStyles-tray-11"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Tray 2
        </p>
        <iframe
          class="makeStyles-iframe-15"
          src="https://flex.twilio.com/admin/"
          title="Twilio Flex"
        />
      </section>
      <section
        class="makeStyles-container-17 makeStyles-panel-18 makeStyles-tray3-12"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Tray 3
        </p>
      </section>
    </section>
  </div>
</DocumentFragment>
`;
